src/audio_unit/looper.rs

```diff
@@ -0,0 +1,137 @@
+use crate::{audio_unit::AudioUnit, util, Result};
+use cpal::StreamConfig;
+use std::sync::mpsc::{self, Receiver, Sender};
+use State::*;
+
+#[derive(Debug)]
+pub enum Message {
+    Toggle,
+}
+
+#[derive(Debug)]
+enum State {
+    Off,
+    Recording { position: usize },
+    Playing { position: usize, total: usize },
+}
+
+pub struct Looper {
+    messages: Receiver<Message>,
+    buffer: Vec<f32>,
+    state: State,
+}
+
+impl Looper {
+    pub fn new(stream_config: &StreamConfig, max_buffer_ms: u32) -> (Self, Sender<Message>) {
+        let (sender, receiver) = mpsc::channel();
+
+        let buffer = vec![0.0; util::ms_in_samples(stream_config, max_buffer_ms)];
+
+        (
+            Self {
+                messages: receiver,
+                buffer,
+                state: Off,
+            },
+            sender,
+        )
+    }
+
+    fn process_messages(&mut self) {
+        let messages: Vec<_> = self.messages.try_iter().collect();
+        for message in messages {
+            self.process_message(message);
+        }
+    }
+
+    fn process_message(&mut self, _: Message) {
+        self.state = match self.state {
+            Off => Recording { position: 0 },
+            Recording { position } => Playing {
+                position: 0,
+                total: position,
+            },
+            Playing {
+                position: _,
+                total: _,
+            } => {
+                util::zero_slice(&mut self.buffer);
```

Comment:
Trying to improve my mental model of the buffers a bit - why do you need to reset the buffer to zero's when turning off the looper?

Reply:
I guess you don't, since to get back into a playback state, you'd have to go through `Recording` which would overwrite everything anyway. I'll update it (after verifying that's true).